[Bug] tsc in a project created with `npx @temporalio/create` fails on Node 20

### What are you really trying to do?

<!-- 
Tell us at a high level what you're doing, to avoid XY problem (https://en.wikipedia.org/wiki/XY_problem) 
-->

Trying to create a new project based on the `child-workflow` sample, following the directions at: https://docs.temporal.io/dev-guide/typescript/foundations#install-a-temporal-sdk

### Describe the bug

<!-- A clear and concise description of what the bug is. -->

<!-- If applicable, add screenshots or code blocks to help explain your problem. You can also use [Loom](http://loom.com/) to do short, free video bug reports. -->

TypeScript compilation fails in an unmodified new project, based on the `child-workflow` sample, because the `es2023` library definitions are not included with the version of TypeScript used in the sample project:

```
❯ npm run start.watch
...
TSError: ⨯ Unable to compile TypeScript:
node_modules/@tsconfig/node20/tsconfig.json:6:13 - error TS6046: Argument for '--lib' option must be: 'es5', 'es6', 'es2015', 'es7', [...]

6     "lib": ["es2023"],
              ~~~~~~~~
...
[nodemon] app crashed - waiting for file changes before starting...
```

### Minimal Reproduction

<!-- 
Modify our hello world templates to demonstrate:

- TypeScript: https://github.com/temporalio/samples-typescript/tree/main/hello-world
- Go: https://github.com/temporalio/money-transfer-project-template-go
- Java: https://github.com/temporalio/money-transfer-project-template-java
- PHP: https://github.com/temporalio/samples-php#samples
-->

1. `npx @temporalio/create@latest sample-tsc-fail-repro`
2. Choose the `child-workflows` sample and accept the defaults for the rest of the prompts
3. In a new terminal, `temporal server start-dev`
4. `cd sample-tsc-fail-repro`
5. `npm run start.watch`
6. Observe the crash above

### Environment/Versions

<!-- Please complete the following information where relevant. -->

- OS and processor: M1 Mac, macOS 13.4
- Temporal Version: `temporal version 0.9.0 (server 1.20.2) (ui 2.15.0)`
- Are you using Docker or Kubernetes or building Temporal from source? No
- Node Version: `v20.2.0`

### Additional context

<!-- Add any other context about the problem here. -->

I doubt this is specific to the `child-workflows` sample; I suspect the others suffer from the same issue (though I haven't tried).

I notice the `es2023` lib reference comes from the `tsconfig.json` included in `@tsconfig/node20`. If I look at the sample itself, I see that its `package.json` uses `@tsconfig/node16` (which presumably does not have the `es2023` lib reference). So it seems `@temporalio/create` is tweaking the `package.json` to use `node20` (which is the version I have installed). And indeed I can see this in the code: https://github.com/temporalio/sdk-typescript/blob/1709d45c5e649fd6b233243b7a80c49be5e706bd/packages/create-project/src/helpers/install.ts#L33

Also, the sample's `package.json` pins TypeScript to a 4.x release, and I don't think `es2023` is defined until 5.0.

I think there might be two ways of fixing this:
1. Bump the TypeScript version (and related dependencies) to TypeScript 5.x in all the samples
2. Stop messing with the `@tsconfig/nodeXX` version when creating the new package

It might be a good idea to do both, in any case. Staying up to date (i.e. (1)) is always good, and my thinking re. (2) is that people may be developing on the latest version of Node (currently, v20), but may want to use an LTS release (currently, v18) when deploying. So it may not necessarily make sense to depend on the current Node version in a new project.

--- a/packages/client/src/schedule-client.ts
+++ b/packages/client/src/schedule-client.ts
@@ -379,7 +379,7 @@ export class ScheduleClient extends BaseClient {
           scheduleId: raw.scheduleId,
 
           spec: decodeScheduleSpec(raw.info?.spec ?? {}),
-          action: raw.info?.workflowType?.name && {
+          action: raw.info?.workflowType && {
             type: 'startWorkflow',
             workflowType: raw.info.workflowType.name,
           },

--- a/packages/client/src/schedule-client.ts
+++ b/packages/client/src/schedule-client.ts
@@ -260,7 +260,6 @@ export class ScheduleClient extends BaseClient {
       const res = await this.workflowService.createSchedule(req);
       return { conflictToken: res.conflictToken };
     } catch (err: any) {
-      if (err instanceof TypeError) throw err;
       if (err.code === grpcStatus.ALREADY_EXISTS) {
         throw new ScheduleAlreadyRunning('Schedule already exists and is running', opts.scheduleId);
       }
@@ -292,21 +291,21 @@ export class ScheduleClient extends BaseClient {
     opts: CompiledScheduleUpdateOptions,
     header: Headers
   ): Promise<temporal.api.workflowservice.v1.IUpdateScheduleResponse> {
+    const req = {
+      namespace: this.options.namespace,
+      scheduleId,
+      schedule: {
+        spec: encodeScheduleSpec(opts.spec),
+        action: await encodeScheduleAction(this.dataConverter, opts.action, header),
+        policies: encodeSchedulePolicies(opts.policies),
+        state: encodeScheduleState(opts.state),
+      },
+      identity: this.options.identity,
+      requestId: uuid4(),
+    };
     try {
-      return await this.workflowService.updateSchedule({
-        namespace: this.options.namespace,
-        scheduleId,
-        schedule: {
-          spec: encodeScheduleSpec(opts.spec),
-          action: await encodeScheduleAction(this.dataConverter, opts.action, header),
-          policies: encodeSchedulePolicies(opts.policies),
-          state: encodeScheduleState(opts.state),
-        },
-        identity: this.options.identity,
-        requestId: uuid4(),
-      });
+      return await this.workflowService.updateSchedule(req);
     } catch (err: any) {
-      if (err instanceof TypeError) throw err;
       this.rethrowGrpcError(err, scheduleId, 'Failed to update schedule');
     }
   }

--- a/packages/client/src/schedule-helpers.ts
+++ b/packages/client/src/schedule-helpers.ts
@@ -2,6 +2,7 @@ import Long from 'long'; // eslint-disable-line import/no-named-as-default
 import {
   compileRetryPolicy,
   decompileRetryPolicy,
+  extractWorkflowType,
   LoadedDataConverter,
   mapFromPayloads,
   mapToPayloads,
@@ -208,7 +209,7 @@ export function decodeOverlapPolicy(input?: temporal.api.enums.v1.ScheduleOverla
 
 export function compileScheduleOptions(options: ScheduleOptions): CompiledScheduleOptions {
   const workflowTypeOrFunc = options.action.workflowType;
-  const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+  const workflowType = extractWorkflowType(workflowTypeOrFunc);
   return {
     ...options,
     action: {
@@ -222,7 +223,7 @@ export function compileScheduleOptions(options: ScheduleOptions): CompiledSchedu
 
 export function compileUpdatedScheduleOptions(options: ScheduleUpdateOptions): CompiledScheduleUpdateOptions {
   const workflowTypeOrFunc = options.action.workflowType;
-  const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+  const workflowType = extractWorkflowType(workflowTypeOrFunc);
   return {
     ...options,
     action: {

--- a/packages/client/src/workflow-client.ts
+++ b/packages/client/src/workflow-client.ts
@@ -18,6 +18,7 @@ import {
   WorkflowExecutionAlreadyStartedError,
   WorkflowNotFoundError,
   WorkflowResultType,
+  extractWorkflowType,
 } from '@temporalio/common';
 import { composeInterceptors } from '@temporalio/common/lib/interceptors';
 import { History } from '@temporalio/common/lib/proto-utils';
@@ -345,7 +346,7 @@ export class WorkflowClient extends BaseClient {
     options: WithWorkflowArgs<T, WorkflowOptions>,
     interceptors: WorkflowClientInterceptor[]
   ): Promise<string> {
-    const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+    const workflowType = extractWorkflowType(workflowTypeOrFunc);
     assertRequiredWorkflowOptions(options);
     const compiledOptions = compileWorkflowOptions(ensureArgs(options));
 
@@ -369,7 +370,7 @@ export class WorkflowClient extends BaseClient {
     options: WithWorkflowArgs<T, WorkflowSignalWithStartOptions<SA>>,
     interceptors: WorkflowClientInterceptor[]
   ): Promise<string> {
-    const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+    const workflowType = extractWorkflowType(workflowTypeOrFunc);
     const { signal, signalArgs, ...rest } = options;
     assertRequiredWorkflowOptions(rest);
     const compiledOptions = compileWorkflowOptions(ensureArgs(rest));

--- a/packages/common/src/workflow-options.ts
+++ b/packages/common/src/workflow-options.ts
@@ -159,7 +159,7 @@ export function extractWorkflowType<T extends Workflow>(workflowTypeOrFunc: stri
   if (typeof workflowTypeOrFunc === 'string') return workflowTypeOrFunc as string;
   if (typeof workflowTypeOrFunc === 'function') {
     if (workflowTypeOrFunc?.name) return workflowTypeOrFunc.name;
-    throw new TypeError(`Invalid workflow type: the workflow function is anonymous`);
+    throw new TypeError('Invalid workflow type: the workflow function is anonymous');
   }
   throw new TypeError(
     `Invalid workflow type: expected either a string or a function, got '${typeof workflowTypeOrFunc}'`

--- a/packages/common/src/workflow-options.ts
+++ b/packages/common/src/workflow-options.ts
@@ -154,3 +154,14 @@ export function compileWorkflowOptions<T extends CommonWorkflowOptions>(options:
     workflowTaskTimeout: msOptionalToTs(workflowTaskTimeout),
   };
 }
+
+export function extractWorkflowType<T extends Workflow>(workflowTypeOrFunc: string | T): string {
+  if (typeof workflowTypeOrFunc === 'string') return workflowTypeOrFunc as string;
+  if (typeof workflowTypeOrFunc === 'function') {
+    if (workflowTypeOrFunc?.name) return workflowTypeOrFunc.name;
+    throw new TypeError(`Invalid workflow type: the workflow function is anonymous`);
+  }
+  throw new TypeError(
+    `Invalid workflow type: expected either a string or a function, got '${typeof workflowTypeOrFunc}'`
+  );
+}

--- a/packages/create-project/src/helpers/install.ts
+++ b/packages/create-project/src/helpers/install.ts
@@ -35,9 +35,8 @@ export async function updateNodeVersion({ root }: InstallArgs): Promise<void> {
   const versionAlreadyInPackageJson = 16;
   const minimumValidVersion = 14;
 
-  // For some reason, the @tsconfig/node20 package exists, but is currently unusable because it
-  // refers to a --lib value that isn't supported in latest release of TypeScript.
-  // FIXME: Update this once @tsconfig/node20 is fixed.
+  // The @tsconfig/node20 sets "--lib es2023", which require TypeScript 5.x.
+  // FIXME: Remove this once samples have been updated to TypeScript ^5.0.0.
   const maximumValidVersion = 18;
 
   const tsconfigNodeVersion = Math.max(minimumValidVersion, Math.min(currentNodeVersion, maximumValidVersion));

--- a/packages/create-project/src/helpers/install.ts
+++ b/packages/create-project/src/helpers/install.ts
@@ -35,8 +35,15 @@ export async function updateNodeVersion({ root }: InstallArgs): Promise<void> {
   const versionAlreadyInPackageJson = 16;
   const minimumValidVersion = 14;
 
-  if (currentNodeVersion >= minimumValidVersion && currentNodeVersion !== versionAlreadyInPackageJson) {
-    const packageName = `@tsconfig/node${currentNodeVersion}`;
+  // For some reason, the @tsconfig/node20 package exists, but is currently unusable because it
+  // refers to a --lib value that isn't supported in latest release of TypeScript.
+  // FIXME: Update this once @tsconfig/node20 is fixed.
+  const maximumValidVersion = 18;
+
+  const tsconfigNodeVersion = Math.max(minimumValidVersion, Math.min(currentNodeVersion, maximumValidVersion));
+
+  if (tsconfigNodeVersion !== versionAlreadyInPackageJson) {
+    const packageName = `@tsconfig/node${tsconfigNodeVersion}`;
 
     const packageExists = await isUrlOk(`https://registry.npmjs.org/${packageName}`);
     if (packageExists) {

--- a/packages/proto/scripts/compile-proto.js
+++ b/packages/proto/scripts/compile-proto.js
@@ -64,10 +64,12 @@ async function compileProtos(dtsOutputFile, ...args) {
     // pbts internally calls jsdoc, which do strict validation of jsdoc tags.
     // Unfortunately, some protobuf comment about cron syntax contains the
     // "@every" shorthand at the begining of a line, making it appear as a
-    // (invalid) jsdoc tag. We fix this by rewriting all cron shorthand
-    // using markdown "code" syntax.
+    // (invalid) jsdoc tag. Similarly, docusaurus trips on <interval> and other
+    // things that looks like html tags. We fix both cases by rewriting these
+    // using markdown "inline code" syntax.
     let tempFileContent = await readFile(tempFile, 'utf8');
     tempFileContent = tempFileContent.replace(/(@(?:yearly|monthly|weekly|daily|hourly|every))/g, '`$1`');
+    tempFileContent = tempFileContent.replace(/<((?:interval|phase|timezone)(?: [^>]+)?)>/g, '`<$1>`');
     await writeFile(tempFile, tempFileContent, 'utf-8');
 
     await promisify(pbts.main)(['--out', dtsOutputFile, tempFile]);

--- a/packages/proto/scripts/compile-proto.js
+++ b/packages/proto/scripts/compile-proto.js
@@ -1,4 +1,4 @@
-const { rm } = require('fs/promises');
+const { rm, readFile, writeFile } = require('fs/promises');
 const { resolve } = require('path');
 const { promisify } = require('util');
 const dedent = require('dedent');
@@ -42,6 +42,7 @@ async function compileProtos(dtsOutputFile, ...args) {
     'commonjs',
     '--force-long',
     '--no-verify',
+    '--alt-comment',
     '--root',
     '__temporal',
     resolve(require.resolve('protobufjs'), '../google/protobuf/descriptor.proto'),
@@ -59,6 +60,16 @@ async function compileProtos(dtsOutputFile, ...args) {
   console.log(`Creating protobuf TS definitions from ${coreProtoPath} and ${workflowServiceProtoPath}`);
   try {
     await promisify(pbjs.main)([...pbjsArgs, '--target', 'static-module', '--out', tempFile]);
+
+    // pbts internally calls jsdoc, which do strict validation of jsdoc tags.
+    // Unfortunately, some protobuf comment about cron syntax contains the
+    // "@every" shorthand at the begining of a line, making it appear as a
+    // (invalid) jsdoc tag. We fix this by rewriting all cron shorthand
+    // using markdown "code" syntax.
+    let tempFileContent = await readFile(tempFile, 'utf8');
+    tempFileContent = tempFileContent.replace(/(@(?:yearly|monthly|weekly|daily|hourly|every))/g, '`$1`');
+    await writeFile(tempFile, tempFileContent, 'utf-8');
+
     await promisify(pbts.main)(['--out', dtsOutputFile, tempFile]);
   } finally {
     await rm(tempFile);

--- a/packages/test/scripts/compile-proto.js
+++ b/packages/test/scripts/compile-proto.js
@@ -28,7 +28,17 @@ function mtime(path) {
 }
 
 async function compileProtos(outputFile, ...args) {
-  const pbjsArgs = [...args, '--wrap', 'commonjs', '--force-long', '--no-verify', '--out', outputFile, ...protoFiles];
+  const pbjsArgs = [
+    ...args,
+    '--wrap',
+    'commonjs',
+    '--force-long',
+    '--no-verify',
+    '--alt-comment',
+    '--out',
+    outputFile,
+    ...protoFiles,
+  ];
   return await promisify(pbjs.main)(pbjsArgs);
 }
 

--- a/packages/workflow/src/workflow.ts
+++ b/packages/workflow/src/workflow.ts
@@ -2,6 +2,7 @@ import {
   ActivityFunction,
   ActivityOptions,
   compileRetryPolicy,
+  extractWorkflowType,
   IllegalStateError,
   LocalActivityOptions,
   mapToPayloads,
@@ -698,7 +699,7 @@ export async function startChild<T extends Workflow>(
 ): Promise<ChildWorkflowHandle<T>> {
   const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
-  const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+  const workflowType = extractWorkflowType(workflowTypeOrFunc);
   const execute = composeInterceptors(
     activator.interceptors.outbound,
     'startChildWorkflowExecution',
@@ -797,7 +798,7 @@ export async function executeChild<T extends Workflow>(
 ): Promise<WorkflowResultType<T>> {
   const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
-  const workflowType = typeof workflowTypeOrFunc === 'string' ? workflowTypeOrFunc : workflowTypeOrFunc.name;
+  const workflowType = extractWorkflowType(workflowTypeOrFunc);
   const execute = composeInterceptors(
     activator.interceptors.outbound,
     'startChildWorkflowExecution',

--- a/packages/workflow/src/workflow.ts
+++ b/packages/workflow/src/workflow.ts
@@ -109,6 +109,7 @@ function timerNextHandler(input: TimerInput) {
  * If given a negative number or 0, value will be set to 1.
  */
 export function sleep(ms: number | string): Promise<void> {
+  assertInWorkflowContext('Workflow.sleep(...) may only be used from a Workflow Execution');
   const activator = getActivator();
   const seq = activator.nextSeqs.timer++;
 
@@ -251,6 +252,7 @@ async function scheduleLocalActivityNextHandler({
  * @hidden
  */
 export function scheduleActivity<R>(activityType: string, args: any[], options: ActivityOptions): Promise<R> {
+  assertInWorkflowContext('Workflow.scheduleActivity(...) may only be used from a Workflow Execution');
   const activator = getActivator();
   if (options === undefined) {
     throw new TypeError('Got empty activity options');
@@ -276,6 +278,7 @@ export async function scheduleLocalActivity<R>(
   args: any[],
   options: LocalActivityOptions
 ): Promise<R> {
+  assertInWorkflowContext('Workflow.scheduleLocalActivity(...) may only be used from a Workflow Execution');
   const activator = getActivator();
   if (options === undefined) {
     throw new TypeError('Got empty activity options');
@@ -575,6 +578,9 @@ const CONDITION_0_PATCH = '__sdk_internal_patch_number:1';
  * It takes a Workflow ID and optional run ID.
  */
 export function getExternalWorkflowHandle(workflowId: string, runId?: string): ExternalWorkflowHandle {
+  assertInWorkflowContext(
+    'Workflow.getExternalWorkflowHandle(...) may only be used from a Workflow Execution. Consider using Client.workflow.getHandle(...) instead.)'
+  );
   const activator = getActivator();
   return {
     workflowId,
@@ -697,6 +703,9 @@ export async function startChild<T extends Workflow>(
   workflowTypeOrFunc: string | T,
   options?: WithWorkflowArgs<T, ChildWorkflowOptions>
 ): Promise<ChildWorkflowHandle<T>> {
+  assertInWorkflowContext(
+    'Workflow.startChild(...) may only be used from a Workflow Execution. Consider using Client.workflow.start(...) instead.)'
+  );
   const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
   const workflowType = extractWorkflowType(workflowTypeOrFunc);
@@ -796,6 +805,9 @@ export async function executeChild<T extends Workflow>(
   workflowTypeOrFunc: string | T,
   options?: WithWorkflowArgs<T, ChildWorkflowOptions>
 ): Promise<WorkflowResultType<T>> {
+  assertInWorkflowContext(
+    'Workflow.executeChild(...) may only be used from a Workflow Execution. Consider using Client.workflow.execute(...) instead.'
+  );
   const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
   const workflowType = extractWorkflowType(workflowTypeOrFunc);
@@ -842,6 +854,7 @@ export async function executeChild<T extends Workflow>(
  * }
  */
 export function workflowInfo(): WorkflowInfo {
+  assertInWorkflowContext('Workflow.workflowInfo(...) may only be used from a Workflow Execution.');
   return getActivator().info;
 }
 
@@ -892,6 +905,7 @@ export function proxySinks<T extends Sinks>(): T {
           {
             get(_, fnName) {
               return (...args: any[]) => {
+                assertInWorkflowContext('Proxied sinks functions may only be used from a Workflow Execution.');
                 const activator = getActivator();
                 activator.sinkCalls.push({
                   ifaceName: ifaceName as string,
@@ -918,6 +932,9 @@ export function proxySinks<T extends Sinks>(): T {
 export function makeContinueAsNewFunc<F extends Workflow>(
   options?: ContinueAsNewOptions
 ): (...args: Parameters<F>) => Promise<never> {
+  assertInWorkflowContext(
+    'Workflow.continueAsNew(...) and Workflow.makeContinueAsNewFunc(...) may only be used from a Workflow Execution.'
+  );
   const activator = getActivator();
   const info = workflowInfo();
   const { workflowType, taskQueue, ...rest } = options ?? {};
@@ -1042,6 +1059,9 @@ export function deprecatePatch(patchId: string): void {
 }
 
 function patchInternal(patchId: string, deprecated: boolean): boolean {
+  assertInWorkflowContext(
+    'Workflow.patch(...) and Workflow.deprecatePatch may only be used from a Workflow Execution.'
+  );
   const activator = getActivator();
   // Patch operation does not support interception at the moment, if it did,
   // this would be the place to start the interception chain
@@ -1076,6 +1096,7 @@ export function condition(fn: () => boolean, timeout: number | string): Promise<
 export function condition(fn: () => boolean): Promise<void>;
 
 export async function condition(fn: () => boolean, timeout?: number | string): Promise<void | boolean> {
+  assertInWorkflowContext('Workflow.condition(...) may only be used from a Workflow Execution.');
   // Prior to 1.5.0, `condition(fn, 0)` was treated as equivalent to `condition(fn, undefined)`
   if (timeout === 0 && !patched(CONDITION_0_PATCH)) {
     return conditionInner(fn);
@@ -1163,6 +1184,7 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
   def: T,
   handler: Handler<Ret, Args, T> | undefined
 ): void {
+  assertInWorkflowContext('Workflow.setHandler(...) may only be used from a Workflow Execution.');
   const activator = getActivator();
   if (def.type === 'signal') {
     if (typeof handler === 'function') {
@@ -1196,6 +1218,7 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
  * @param handler a function that will handle signals for non-registered signal names, or `undefined` to unset the handler.
  */
 export function setDefaultSignalHandler(handler: DefaultSignalHandler | undefined): void {
+  assertInWorkflowContext('Workflow.setDefaultSignalHandler(...) may only be used from a Workflow Execution.');
   const activator = getActivator();
   if (typeof handler === 'function') {
     activator.defaultSignalHandler = handler;
@@ -1237,6 +1260,7 @@ export function setDefaultSignalHandler(handler: DefaultSignalHandler | undefine
  * @param searchAttributes The Record to merge. Use a value of `[]` to clear a Search Attribute.
  */
 export function upsertSearchAttributes(searchAttributes: SearchAttributes): void {
+  assertInWorkflowContext('Workflow.upsertSearchAttributes(...) may only be used from a Workflow Execution.');
   const activator = getActivator();
 
   const mergedSearchAttributes = { ...activator.info.searchAttributes, ...searchAttributes };
@@ -1282,6 +1306,7 @@ export const log: LoggerSinks['defaultWorkerLogger'] = Object.fromEntries(
     return [
       level,
       (message: string, attrs: Record<string, unknown>) => {
+        assertInWorkflowContext('Workflow.log(...) may only be used from a Workflow Execution.)');
         return loggerSinks.defaultWorkerLogger[level](message, {
           // Inject the call time in nanosecond resolution as expected by the worker logger.
           [LogTimestamp]: getActivator().getTimeOfDay(),
@@ -1291,3 +1316,9 @@ export const log: LoggerSinks['defaultWorkerLogger'] = Object.fromEntries(
     ];
   })
 ) as any;
+
+function assertInWorkflowContext(message: string): void {
+  if (maybeGetActivator() == null) {
+    throw new IllegalStateError(message);
+  }
+}

--- a/packages/workflow/src/workflow.ts
+++ b/packages/workflow/src/workflow.ts
@@ -38,7 +38,7 @@ import {
   Handler,
   WorkflowInfo,
 } from './interfaces';
-import { LocalActivityDoBackoff, getActivator, maybeGetActivator } from './internals';
+import { Activator, LocalActivityDoBackoff, getActivator, maybeGetActivator } from './internals';
 import { LoggerSinks, Sinks } from './sinks';
 import { untrackPromise } from './stack-helpers';
 import { ChildWorkflowHandle, ExternalWorkflowHandle } from './workflow-handle';
@@ -109,8 +109,7 @@ function timerNextHandler(input: TimerInput) {
  * If given a negative number or 0, value will be set to 1.
  */
 export function sleep(ms: number | string): Promise<void> {
-  assertInWorkflowContext('Workflow.sleep(...) may only be used from a Workflow Execution');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext('Workflow.sleep(...) may only be used from a Workflow Execution');
   const seq = activator.nextSeqs.timer++;
 
   const durationMs = Math.max(1, msToNumber(ms));
@@ -252,8 +251,9 @@ async function scheduleLocalActivityNextHandler({
  * @hidden
  */
 export function scheduleActivity<R>(activityType: string, args: any[], options: ActivityOptions): Promise<R> {
-  assertInWorkflowContext('Workflow.scheduleActivity(...) may only be used from a Workflow Execution');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext(
+    'Workflow.scheduleActivity(...) may only be used from a Workflow Execution'
+  );
   if (options === undefined) {
     throw new TypeError('Got empty activity options');
   }
@@ -278,8 +278,9 @@ export async function scheduleLocalActivity<R>(
   args: any[],
   options: LocalActivityOptions
 ): Promise<R> {
-  assertInWorkflowContext('Workflow.scheduleLocalActivity(...) may only be used from a Workflow Execution');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext(
+    'Workflow.scheduleLocalActivity(...) may only be used from a Workflow Execution'
+  );
   if (options === undefined) {
     throw new TypeError('Got empty activity options');
   }
@@ -578,10 +579,9 @@ const CONDITION_0_PATCH = '__sdk_internal_patch_number:1';
  * It takes a Workflow ID and optional run ID.
  */
 export function getExternalWorkflowHandle(workflowId: string, runId?: string): ExternalWorkflowHandle {
-  assertInWorkflowContext(
+  const activator = assertInWorkflowContext(
     'Workflow.getExternalWorkflowHandle(...) may only be used from a Workflow Execution. Consider using Client.workflow.getHandle(...) instead.)'
   );
-  const activator = getActivator();
   return {
     workflowId,
     runId,
@@ -703,10 +703,9 @@ export async function startChild<T extends Workflow>(
   workflowTypeOrFunc: string | T,
   options?: WithWorkflowArgs<T, ChildWorkflowOptions>
 ): Promise<ChildWorkflowHandle<T>> {
-  assertInWorkflowContext(
+  const activator = assertInWorkflowContext(
     'Workflow.startChild(...) may only be used from a Workflow Execution. Consider using Client.workflow.start(...) instead.)'
   );
-  const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
   const workflowType = extractWorkflowType(workflowTypeOrFunc);
   const execute = composeInterceptors(
@@ -805,10 +804,9 @@ export async function executeChild<T extends Workflow>(
   workflowTypeOrFunc: string | T,
   options?: WithWorkflowArgs<T, ChildWorkflowOptions>
 ): Promise<WorkflowResultType<T>> {
-  assertInWorkflowContext(
+  const activator = assertInWorkflowContext(
     'Workflow.executeChild(...) may only be used from a Workflow Execution. Consider using Client.workflow.execute(...) instead.'
   );
-  const activator = getActivator();
   const optionsWithDefaults = addDefaultWorkflowOptions(options ?? ({} as any));
   const workflowType = extractWorkflowType(workflowTypeOrFunc);
   const execute = composeInterceptors(
@@ -854,8 +852,8 @@ export async function executeChild<T extends Workflow>(
  * }
  */
 export function workflowInfo(): WorkflowInfo {
-  assertInWorkflowContext('Workflow.workflowInfo(...) may only be used from a Workflow Execution.');
-  return getActivator().info;
+  const activator = assertInWorkflowContext('Workflow.workflowInfo(...) may only be used from a Workflow Execution.');
+  return activator.info;
 }
 
 /**
@@ -905,8 +903,9 @@ export function proxySinks<T extends Sinks>(): T {
           {
             get(_, fnName) {
               return (...args: any[]) => {
-                assertInWorkflowContext('Proxied sinks functions may only be used from a Workflow Execution.');
-                const activator = getActivator();
+                const activator = assertInWorkflowContext(
+                  'Proxied sinks functions may only be used from a Workflow Execution.'
+                );
                 activator.sinkCalls.push({
                   ifaceName: ifaceName as string,
                   fnName: fnName as string,
@@ -932,11 +931,10 @@ export function proxySinks<T extends Sinks>(): T {
 export function makeContinueAsNewFunc<F extends Workflow>(
   options?: ContinueAsNewOptions
 ): (...args: Parameters<F>) => Promise<never> {
-  assertInWorkflowContext(
+  const activator = assertInWorkflowContext(
     'Workflow.continueAsNew(...) and Workflow.makeContinueAsNewFunc(...) may only be used from a Workflow Execution.'
   );
-  const activator = getActivator();
-  const info = workflowInfo();
+  const info = activator.info;
   const { workflowType, taskQueue, ...rest } = options ?? {};
   const requiredOptions = {
     workflowType: workflowType ?? info.workflowType,
@@ -1059,10 +1057,9 @@ export function deprecatePatch(patchId: string): void {
 }
 
 function patchInternal(patchId: string, deprecated: boolean): boolean {
-  assertInWorkflowContext(
+  const activator = assertInWorkflowContext(
     'Workflow.patch(...) and Workflow.deprecatePatch may only be used from a Workflow Execution.'
   );
-  const activator = getActivator();
   // Patch operation does not support interception at the moment, if it did,
   // this would be the place to start the interception chain
 
@@ -1184,8 +1181,7 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
   def: T,
   handler: Handler<Ret, Args, T> | undefined
 ): void {
-  assertInWorkflowContext('Workflow.setHandler(...) may only be used from a Workflow Execution.');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext('Workflow.setHandler(...) may only be used from a Workflow Execution.');
   if (def.type === 'signal') {
     if (typeof handler === 'function') {
       activator.signalHandlers.set(def.name, handler as any);
@@ -1218,8 +1214,9 @@ export function setHandler<Ret, Args extends any[], T extends SignalDefinition<A
  * @param handler a function that will handle signals for non-registered signal names, or `undefined` to unset the handler.
  */
 export function setDefaultSignalHandler(handler: DefaultSignalHandler | undefined): void {
-  assertInWorkflowContext('Workflow.setDefaultSignalHandler(...) may only be used from a Workflow Execution.');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext(
+    'Workflow.setDefaultSignalHandler(...) may only be used from a Workflow Execution.'
+  );
   if (typeof handler === 'function') {
     activator.defaultSignalHandler = handler;
     activator.dispatchBufferedSignals();
@@ -1260,8 +1257,9 @@ export function setDefaultSignalHandler(handler: DefaultSignalHandler | undefine
  * @param searchAttributes The Record to merge. Use a value of `[]` to clear a Search Attribute.
  */
 export function upsertSearchAttributes(searchAttributes: SearchAttributes): void {
-  assertInWorkflowContext('Workflow.upsertSearchAttributes(...) may only be used from a Workflow Execution.');
-  const activator = getActivator();
+  const activator = assertInWorkflowContext(
+    'Workflow.upsertSearchAttributes(...) may only be used from a Workflow Execution.'
+  );
 
   const mergedSearchAttributes = { ...activator.info.searchAttributes, ...searchAttributes };
   if (!mergedSearchAttributes) {
@@ -1317,8 +1315,8 @@ export const log: LoggerSinks['defaultWorkerLogger'] = Object.fromEntries(
   })
 ) as any;
 
-function assertInWorkflowContext(message: string): void {
-  if (maybeGetActivator() == null) {
-    throw new IllegalStateError(message);
-  }
+function assertInWorkflowContext(message: string): Activator {
+  const activator = maybeGetActivator();
+  if (activator == null) throw new IllegalStateError(message);
+  return activator;
 }

